design-documents/promises.md

```diff
@@ -0,0 +1,240 @@
+ 
+ 
+##### Callback hell
+Consider this code responsible for placing orders
+```php
+class ServiceA
+{
+....
+
+    public function process(DTOInterface $dto): ProcessedInterface
+    {
+        ....
+        
+        $this->serviceB->processSmth($val)->then(function ($val) use ($processed) {
+            $processed->setBValue($val);
+        });
+        
+        return $processed;
+    }
+
+....
+}
+```
+We cannot be sure _serviceB_ has finished doing it's stuff when we return _$processed_.
+So, if we cannot wait for the promise _serviceB_ returned, the only thing we can do
+is to return a promise ourselves instead of _ProcessedInterface_. But then methods using
+_ServiceA::process()_ would have to do the same - and PHP code is not supposed to be this way.
+
+##### Wait not unwrapping promises
+For _wait_ method to also unwrap promises can result in confusion. It's better to have a single way of retreiving promised results and a single way of retreiving errors.
+ 
+Consider next situation:
+```php
+class ServiceA
+{
+    public function doSmth(): void
+    {
+       $promise = $this->otherService->doSmthElse();
+       $promise->otherwise(function ($exception) { $this->logger->critical($exception); });
+       try {
+           //wait would throw the exception processed in the otherwise callback once again
+           $promise->wait();
+       } catch (\Throwable $exception) {
+           //we've already processed this
+       }
+    }
+}
+```
+That is a simple situation but it illustrates how having multiple ways of receiving promised results may lead to duplicating code
+
+### Using promises for service contracts
+#### Why use promises for service contracts?
+Another way that was proposed to execute service contracts in an asynchronous manner was to use async web API, but there
+are number of problems with that approach:
+* Async web API allows execution of the same operation with different sets of arguments, but not different operations
+* Async web API was meant for execution big number of operations at the same time (thouthands) which is not the case
+  for most functionality and mostly fits only import
+* Since only 1 type of operation can be executed at the same time it will be impossible to execute service contracts
+  from different domains at the same time
+* Async web API uses status requests to check whether operations are completed which is alright for large numbers
+  of operations but for a small number (like 2, 3) it will just generate more requests than just sending 1 request
+  for each operation
+
+So to allow execution of multiple service contracts from different domains it's best to send 1 request per operation
+and to let client code to chain, pass and properly receive promises of results of operations.
+ 
+#### How will it look?
+There are to ways we can go about using promises for asynchronous execution of service contracts:
+* Service interfaces themselves returning promises for client code to use
```

Comment:
How will this work in the existing system? In web API for instance?

Reply:
web API controllers would just have to updated to expect promises, calling wait on them and then rendering results